### PR TITLE
bevy_reflect: Ignored field order

### LINKED ISSUE #5101: bevy_reflect: ignored fields are not ignored by `FromReflect`
**bevy version**:  main as of 1bd33ca

When deriving `FromReflect` on tuple structs, the field index of fields declared after a field marked with `#[reflect(ignore)]` are not as one would expect them.

Here is an example:
```rust
#[derive(Reflect, FromReflect)]
struct Foo(
  u32, // :0
  #[reflect(ignore)]
  f32, // :1
  String, // :2
);

let mut tuple_struct = DynamicTupleStruct::default();
tuple_struct.set_name("Foo".to_owned());
tuple_struct.insert(22_u32);
tuple_struct.insert(3432423234234_i128);
tuple_struct.insert("hello world".to_owned());

let foo = Foo::from_reflect(&tuple_struct).unwrap();

println!(
    "tuple_struct\n[0]: {:?}\n[1]: {:?}\n[2]: {:?}\n",
    tuple_struct.field(0),
    tuple_struct.field(1),
    tuple_struct.field(2),
);
println!(
    "foo\n[0]: {:?}\n[1]: {:?}\n[2]: {:?}\n",
    foo.field(0),
    foo.field(1),
    foo.field(2).
);
```

Prints the following:
```
tuple_struct
[0]: Some(22)
[1]: Some(3432423234234)
[2]: Some("hello world")

foo
[0]: Some(22)
[1]: Some("hello world")
[2]: None
```

The unexpected behavior is here:
```rust
let mut tuple_struct = DynamicTupleStruct::default();
tuple_struct.set_name("Foo".to_owned());
tuple_struct.insert(22_u32);
// vvv required, otherwise Foo::from_reflect returns None
tuple_struct.insert(3432423234234_i128);
tuple_struct.insert("hello world".to_owned());

let foo = Foo::from_reflect(&tuple_struct).unwrap();
```

You'd expect to be able to build `Foo` from a `DynamicTupleStruct` with exactly **two** fields, not three. The correct behavior would be to account for it in the `FromReflect` implementation.

### PR DESCRIPTION
**Pull request by MrGVSV**
# Objective

Fixes #5101

## Solution

Make it so that placing `#[reflect(ignore)]` or `#[reflect(skip_serializing)]` on a field that is not at the end of the type definition results in a compile error.

#### Rationale

Invalid ordering is a common and annoying footgun that can be very difficult to debug— or at least very frustrating. Ignoring a field in the middle of your struct can throw off the index you get back from `Struct::index_of` and friends. This is made even worse for things like tuple structs, where doing so breaks both `FromReflect` and serialization.

It should not be so easy to completely break reflection when using the derive macro. Thus, I think it warrants an error so we prevent users from entering this state in the first place.

#### Alternatives

One alternative, would be to output a compiler warning. Unfortunately, that API is still [unstable](https://github.com/rust-lang/rust/issues/54140).

We could also print out a warning at runtime whenever we rely on ordering. However, as discussed in [this comment](https://github.com/bevyengine/bevy/issues/5101#issuecomment-1281882546), we run the risk of some things not being caught (namely by third-party crates). Doing this is also just more cumbersome and prone to error.

Lastly, we could potentially store some sort of mapping that gives us the correct _declaration_ index for a given index. This might work, but makes things like `FromReflect` a bit more difficult to implement. It also means we need to add a public API to access this information, which may be confusing/unintuitive. That being said, this is probably the most viable alternative out of the three.

---

## Changelog

- `#[reflect(ignore)]` and `#[reflect(skip_serializing)]` attributes now result in a compile error if placed in an invalid order

## Migration Guide

Fields marked with `#[reflect(ignore)]` must now be placed at the end of their respective type definition. This applies to all structs, tuple structs, and enum variants.

```rust
// OLD:
#[derive(Reflect)]
struct MyStruct {
  #[reflect(ignore)] // <- ERROR!
  _ignore_me: i32,
  foo: i32,
  bar: i32,
}

// NEW:
#[derive(Reflect)]
struct MyStruct {
  foo: i32,
  bar: i32,
  #[reflect(ignore)] // <- OK!
  _ignore_me: i32,
}
```

Additionally, `#[reflect(skip_serializing)]` must also be placed at the end of the type definition. However, it must also come _before_ `#[reflect(ignore)]`.

```rust
// OLD:
#[derive(Reflect)]
struct MyStruct {
  #[reflect(ignore)] // <- ERROR!
  _ignore_me: i32,
  #[reflect(skip_serializing)] // <-ERROR!
  _skip_me: i32,
  foo: i32,
}

// NEW:
#[derive(Reflect)]
struct MyStruct {
  foo: i32,
  #[reflect(skip_serializing)] // <- OK!
  _skip_me: i32,
  #[reflect(ignore)] // <- OK!
  _ignore_me: i32,
}
```

### CODE REVIEW

**Review by soqb — APPROVED** [review]

**Review by nicopap — APPROVED** [review]
LGTM, would be better with compilation failure test on `ignore` attribute used not-at-the-end.

File: crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:220]: Yeah I think I originally had it more specific to parsing fields so it made a _little_ more sense then. But now it's since been simplified down to the very basics. Maybe I'll move it to the `utility` module so it could be used elsewhere.

File: crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:170]: Hm, yeah I'm not sure how much we gain from that if we need to then duplicate the arm to handle the error:
```rust
// Handle good value:
Meta::NameValue(MetaNameValue { lit: Lit::Str(lit_str), path, .. } if path.is…
// Handle bad value:
Meta::NameValue(MetaNameValue { lit, path, .. } if path.is…
```

File: crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs]: Yeah I think I like that better haha

**Comment by MrGVSV** [comment]
Blocked on #7041. Moving to draft until that PR is merged.

**Comment by MrGVSV** [comment]
This should be good to go now!

File: crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs
Comment (cart) [inline crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:193]: This is a logical fix to a real problem! Users deriving reflect like this already have control over their field orders, so asking them to change the order in these cases seems reasonable in _most_ cases. However, theres always the chance that field order matters for some other reason (ex: the type is also used for binary serialization: RPC, bincode, GPU types, etc).

Given that Reflect aims to be a "general purpose reflection library", I think breaking these cases will eventually be an issue, both in Bevy projects and elsewhere.

Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:193]: Fair enough. Would it be better to put this behind a default feature? Or at least provide an opt-out attribute like `#[reflect(allow_ignore_order)]` (or some other better name lol)?

Comment (cart) [reply crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:193]: If opting out breaks the scenarios we're trying to fix in this PR (FromReflect and serialization), and this PR breaks other scenarios (reflecting structs that needs specific field orders), I think thats an indicator that we need to rethink our implementation generally / come up with a solution that handles all of these scenarios correctly.

Specifically: is it possible for us to support cases where reflected field indices don't line up exactly with declared field indices. Or alternatively, can we adjust our reflection apis to account for "non-existent / skipped" fields (ensuring reflected field indices always match declared field indices). If either of these are possible, what are the costs?  Performance? Ergonomics? Do we need to mitigate those costs?

Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:193]: > Specifically: is it possible for us to support cases where reflected field indices don't line up exactly with declared field indices. Or alternatively, can we adjust our reflection apis to account for "non-existent / skipped" fields (ensuring reflected field indices always match declared field indices).

I just put up an alternative PR (#7575) which is more in line with this. The only downside is that it requires fields with `#[reflect(skip_serializing)]` to provide some way of generating a default instance (either with a `Default` impl or with a custom function). As noted in the PR, though, I think it makes sense to add such a requirement.

Comment (cart) [reply crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:193]: Brilliant!

**Comment by MrGVSV** [comment]
Closing in favor of #7575

**Pull request closed without merging**
